packages/amplify-prompts/src/progressbars/progressbar.ts

```diff
@@ -0,0 +1,252 @@
+    private total: number;
+    private terminal: Terminal;
+    private payload!: ProgressPayload;
+    private isActive: boolean;
+    private options: BarOptions;
+    private lastDrawnTime: number;
+    private items: Item[];
+
+    barCompleteString: string;
+    barIncompleteString: string;
+    barSize: number;
+
+    constructor(options: BarOptions) {
+      this.terminal = new Terminal();
+      this.value = 0;
+      this.total = 1;
+      this.options = options;
+      this.items = [];
+      this.isActive = false;
+      this.lastDrawnTime = 0;
+      this.barSize = options.barSize || 40;
+
+      this.barCompleteString = (new Array(this.barSize + 1).join(options.barCompleteChar || '='));
+      this.barIncompleteString = (new Array(this.barSize + 1).join(options.barIncompleteChar || '-'));
+    }
+
+    /**
+     * Create bar strings according to progress using the complete and incomplete bar string.
+     */
+    createBarString() : string {
+      const completeSize = Math.round((this.value / this.total) * 40);
+      const incompleteSize = this.barSize - completeSize;
+
+      // generate bar string by stripping the pre-rendered strings
+      const bar = this.barCompleteString.slice(0, completeSize)
+                + this.barIncompleteString.slice(0, incompleteSize);
+
+      return ` || ${bar} || ${this.value}/${this.total}`;
+    }
+
+    /**
+     * Render strings are made by concatenating the progress bar strings with the item strings.
+     */
+    getRenderStrings() : string[] {
+      let finalStrings = [];
+      const progressBar = this.options.progressBarFormatter.call(this, this.payload) + this.createBarString();
+      finalStrings.push(progressBar);
+      finalStrings = this.items.reduce((prev, _current) => prev.concat(`\t${_current.renderString}`), finalStrings);
+      return finalStrings;
+    }
+
+    /**
+     * Return current value.
+     */
+    getValue() : number {
+      return this.value;
+    }
+
+    /**
+     * Render function
+     */
+    render() : void {
+      const stringsToRender = this.getRenderStrings();
+      this.terminal.writeLines(stringsToRender);
+    }
+
+    /**
+     * Checks if progress bar finished.
+     */
+    hasFinished() : boolean {
+      return this.value === this.total;
+    }
+
+    /**
+     * Checks if progress bar failed.
+     */
+    hasFailed() : boolean {
+      let flag = false;
+      this.items.forEach(item => {
+        if (this.options.itemFailedStatus.includes(item.status)) {
+          flag = true;
+        }
+      });
+      return flag;
+    }
+
+    /**
+     * Checks if the environment is tty
+     */
+    isTTY() : boolean {
+      return this.terminal.isTTY();
+    }
+
+    /**
+     * Starts a progress bar and calls render function.
+     */
+    start(total: number, startValue: number, payload: ProgressPayload) : void {
+      this.value = startValue || 0;
+      this.total = (typeof total !== 'undefined' && total >= 0) ? total : 100;
```

Comment:
When would this condition be useful? I'm wondering if it would make more sense to throw an error if this function is called with an invalid value for total.